BaseXX Support and Refactor (#5216)

BigFish2086 opened the pull request:
- [X] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository.
- [X] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style).
- [X] I've documented every `RZ_API` function and struct this PR changes.
- [X] I've added tests that prove my changes are effective (required for changes to `RZ_API`).
- [ ] I've updated the [Rizin book](https://github.com/rizinorg/book) with the relevant information (if needed).

**Detailed description**

This PR addresses the enhancements & features requested in #5216:

- [X] Renamed `librz/util/ubase64.c` → `librz/util/base64.c` for naming consistency.

- [X] Base85
  - Refactored `base85` implementation: Replaced legacy FILE-based API with a buffer-based API similar to `base64`, improving testability and integration (e.g., with `rz-hash`).
  - Exposed `base85` functionality in the `rz-hash` binary.
  - Added Doxygen documentation for `base85.c`.
  - Implemented unit tests for `base85` encoding and decoding.

- [X] Base36
  - Extracted base36 logic from `subprojects/rzwinkd/iob_net.c` into a standalone `librz/util/base36.c` as suggested.
  - Change the return value on error from 0 to -1 to be easier when testing for errors.
  - Implemented encoding support alongside the existing decoding function.
  - Adopted a unified API similar to `base64`, and exposed `base36` in `rz-hash`.
  - Added Doxygen documentation for `base36.c`.
  - Implemented unit tests for `base36`.

- [X] Base32
  - Implemented `base32` encoding and decoding functions.
  - Exposed `base32` in the `rz-hash` binary.
  - Added Doxygen documentation for `base32.c`.
  - Implemented unit tests for `base32`.

- [X] Base16
  - Implemented `base16` encoding and decoding functions.
  - Exposed `base16` in the `rz-hash` binary.
  - Added Doxygen documentation for `base16.c`.
  - Implemented unit tests for `base16`.

**Test plan**

- this PR should be tested via unit-tests like this: `meson test -C build base85 base64 base36 base32 base16`
- also, the rz-hash binary extensions should be tested, like testing encoding & decoding some data using base85, base64, base36, base32 and base16
- also, one should make sure that new algorithms are being listed correctly in `rz-hash -L`

**Closing issues**

closes #5216

BigFish2086 committed:
Rename librz/util/ubase64.c into librz/util/base64.c for consistency #5216

BigFish2086 committed:
Add Doxygen documentation to function in librz/util/base85.c #5216

BigFish2086 committed:
Add unit-tests for librz/util/base85.c #5216

BigFish2086 committed:
Refactor base85 API and internal functions #5216

base85 was using a FILE as an input and it used to print its result to
STDOUT, which was very bad for having unit-testing or to use it to
extend rz-hash functionality, since the solution I really had was
through using pipes which isn't ideal as much as the refactoring one.

BigFish2086 committed:
Extend rz-hash with base85 #5126

BigFish2086 committed:
Fix formatting #5216

BigFish2086 committed:
Extract base36 decoding function from subprojects/rzwinkd/iob_net.c to librz/util/base36.c #5216

- Also, added Doxygen docs to rz_base36_decode function

BigFish2086 committed:
Added rz_base36_encode_dyn #5216

- this function encode a u64 value to it's char[13+1] base36 counterpart
- later, it will be helpful in expanding the funcitionality of rz-hash
- added its Doxygen docs

BigFish2086 committed:
Expand rz-hash with base36 encoding/decoding #5216

BigFish2086 committed:
Refactor rz_base36_decode to return a suitable error code like -1 #5216

- that should help testing it later for valid and invalid decodings
- keeping the original behaviour as it was in
subprojects/rzwinkd/iob_net.c

BigFish2086 committed:
Add unit-tests for base36 encoding & decoding functions #5216

BigFish2086 committed:
Fix formatting #5216

BigFish2086 committed:
Fix base36.h header file guards #5216

BigFish2086 committed:
Add base32 encoding & decoding #5216

added their Doxygen Docs as well

BigFish2086 committed:
Add base32 encoding and decoding unti-tests #5216

BigFish2086 committed:
Expand rz-hash with base32 encoding/decoding functionality #5216

BigFish2086 committed:
Fix formatting #5216

BigFish2086 committed:
Fix add base32 to codec_name_bytes #5216

BigFish2086 committed:
Fix base32 docs #5216

BigFish2086 committed:
Add base16 encoding/decoding functions #5216

- added their Doxygen docs as well.

BigFish2086 committed:
Add base16 unit-tests #5216

BigFish2086 committed:
Fix test_base85 conversions warnings #5216

BigFish2086 committed:
Expand rz-hash with base16 encoding/decoding functionality #5216

BigFish2086 committed:
Fix formatting #5216

BigFish2086 committed:
Fix doxygen docs #5216

Rot127 commented:
Thanks for the contribution!

ASAN detects some overflows. You can see our `BUILDING.md` how to build with ASAN locally.

```
19/111 rizin:unit / base85               FAIL            0.45s   exit status 1
>>> LD_LIBRARY_PATH=/home/runner/work/rizin/rizin/build/librz/reg:/home/runner/work/rizin/rizin/build/librz/util:/home/runner/work/rizin/rizin/build/librz/lang:/home/runner/work/rizin/rizin/build/librz/config:/home/runner/work/rizin/rizin/build/librz/sign:/home/runner/work/rizin/rizin/build/librz/egg:/home/runner/work/rizin/rizin/build/librz/core:/home/runner/work/rizin/rizin/build/librz/main:/home/runner/work/rizin/rizin/build/librz/arch:/home/runner/work/rizin/rizin/build/librz/magic:/home/runner/work/rizin/rizin/build/librz/type:/home/runner/work/rizin/rizin/build/librz/bin:/home/runner/work/rizin/rizin/build/librz/il:/home/runner/work/rizin/rizin/build/librz/cons:/home/runner/work/rizin/rizin/build/librz/flag:/home/runner/work/rizin/rizin/build/librz/hash:/home/runner/work/rizin/rizin/build/librz/crypto:/home/runner/work/rizin/rizin/build/librz/io:/home/runner/work/rizin/rizin/build/librz/diff:/home/runner/work/rizin/rizin/build/librz/demangler:/home/runner/work/rizin/rizin/build/librz/search:/home/runner/work/rizin/rizin/build/librz/socket:/home/runner/work/rizin/rizin/build/librz/debug:/home/runner/work/rizin/rizin/build/librz/syscall:/home/runner/lib/x86_64-linux-gnu:/home/runner/lib:/home/runner/lib64: UBSAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1 MESON_TEST_ITERATION=1 MSAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1 MALLOC_PERTURB_=71 /home/runner/work/rizin/rizin/build/test/unit/test_base85
――――――――――――――――――――――――――――――――――――― ✀  ―――――――――――――――――――――――――――――――――――――
stderr:
=================================================================
==15411==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x50200000001d at pc 0x7f2dcd307e1b bp 0x7ffdb1764160 sp 0x7ffdb1764150
WRITE of size 1 at 0x50200000001d thread T0
    #0 0x7f2dcd307e1a in decode_tuple_buf ../librz/util/base85.c:186
    #1 0x7f2dcd309415 in rz_base85_decode ../librz/util/base85.c:386
    #2 0x7f2dcd309765 in rz_base85_decode_dyn ../librz/util/base85.c:416
    #3 0x556001d8ef67 in test_rz_base85_decode_dyn_nodelims ../test/unit/test_base85.c:29
    #4 0x556001d915a5 in all_tests ../test/unit/test_base85.c:112
    #5 0x7f2dcbe2a1c9  (/lib/x86_64-linux-gnu/libc.so.6+0x2a1c9) (BuildId: 282c2c16e7b6600b0b22ea0c99010d2795752b5f)
    #6 0x7f2dcbe2a28a in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2a28a) (BuildId: 282c2c16e7b6600b0b22ea0c99010d2795752b5f)
    #7 0x556001d8e4e4 in _start (/home/runner/work/rizin/rizin/build/test/unit/test_base85+0x34e4) (BuildId: 06d80149a451ff7e49a65991548ecfe55c06b360)

0x50200000001d is located 0 bytes after 13-byte region [0x502000000010,0x50200000001d)
allocated by thread T0 here:
    #0 0x7f2dce2fd9c7 in malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:69
    #1 0x7f2dcd309741 in rz_base85_decode_dyn ../librz/util/base85.c:411
    #2 0x556001d8ef67 in test_rz_base85_decode_dyn_nodelims ../test/unit/test_base85.c:29
    #3 0x556001d915a5 in all_tests ../test/unit/test_base85.c:112
    #4 0x7f2dcbe2a1c9  (/lib/x86_64-linux-gnu/libc.so.6+0x2a1c9) (BuildId: 282c2c16e7b6600b0b22ea0c99010d2795752b5f)
    #5 0x7f2dcbe2a28a in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2a28a) (BuildId: 282c2c16e7b6600b0b22ea0c99010d2795752b5f)
    #6 0x556001d8e4e4 in _start (/home/runner/work/rizin/rizin/build/test/unit/test_base85+0x34e4) (BuildId: 06d80149a451ff7e49a65991548ecfe55c06b360)

SUMMARY: AddressSanitizer: heap-buffer-overflow ../librz/util/base85.c:186 in decode_tuple_buf
Shadow bytes around the buggy address:
  0x501ffffffd80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x501ffffffe00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x501ffffffe80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x501fffffff00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x501fffffff80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x502000000000: fa fa 00[05]fa fa fa fa fa fa fa fa fa fa fa fa
  0x502000000080: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x502000000100: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x502000000180: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x502000000200: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x502000000280: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==15411==ABORTING
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

```

wargio commented on librz/util/base16.c:
```diff
@@ -0,0 +1,290 @@
+			return -1;
+		}
+		len = (st64)real_len;
+	}
```
invert the if and return early

BigFish2086 replied:
updated in 3ef4d5ceb9

wargio commented on librz/util/base16.c:
```diff
@@ -0,0 +1,290 @@
+	ut8 group[1];
+	size_t ret;
+	rz_return_val_if_fail(src, 0);
+	rz_return_val_if_fail(dest, 0);
```
```suggestion
	rz_return_val_if_fail(src && dest, 0);
```
And move this at line 127

BigFish2086 replied:
updated in 9a4203e4c7

wargio commented on librz/util/base16.c:
```diff
@@ -0,0 +1,290 @@
+ * If \p dest or \p src is \c NULL, nothing is written and \c 0 is returned.
+ */
+RZ_API size_t rz_base16_encode(RZ_OUT RZ_NULLABLE char *dest, RZ_NULLABLE const ut8 *src, size_t n) {
+	ut8 group[1];
+	size_t ret;
```
initialize the vars.

BigFish2086 replied:
updated in 1a0d24d647

wargio commented on librz/util/base16.c:
```diff
@@ -0,0 +1,290 @@
+		return NULL;
+	}
+	(void)rz_base16_encode(out, src, n);
+	return out;
```
i don't see where `out[buf_sz-1]` is set to zero. malloc returns uninitialized memory.

BigFish2086 replied:
updated in 5cc91101b5

wargio commented on librz/util/base16.c:
```diff
@@ -0,0 +1,290 @@
+
+	for (i = j = 0; i < (size_t)n; i++) {
+		int c = src[i];
+		if (is_base16(c)) {
+			buf[j++] = c;
+			if (j == 2) {
+				unpack_from4(tmp, (const ut8 *)buf);
+				*dest++ = tmp[0];
+				ret += 1;
+				j = 0;
+			}
+		} else {
+			return -1;
+		}
```
invert and early return.

BigFish2086 replied:
updated in 86673eac1afb874fb7f24d4944c15ef9183e9732

wargio commented on librz/util/base16.c:
```diff
@@ -0,0 +1,290 @@
+	rz_return_val_if_fail(src, NULL);
+
+	len = calculate_src_length(src, len);
+	if (len < 0 || (len % 2) != 0) {
```
```suggestion
	if (len < 0 || len & 1) {
```

What happens if len is zero?

BigFish2086 replied:
update in 4833c01ef6

wargio commented on librz/util/base16.c:
```diff
@@ -0,0 +1,290 @@
+	/* NUL‑terminate and shrink to fit. */
+	buf[out_len] = '\0';
+	if (out_len + 1 < cap) {
+		ut8 *tmp = (ut8 *)realloc(buf, (size_t)out_len + 1);
+		if (tmp) {
+			buf = tmp;
+		}
```
if `realloc` returns `NULL`, probably we should free `buf` and return `NULL`.

BigFish2086 replied:
I think, We don't need to free buf when realloc fails because we’re shrinking the buffer and the original buf remains valid. If realloc returns NULL, we just keep using the original allocation as-is. Also, this is also how it is in case of `base64.c` for example:

https://github.com/rizinorg/rizin/blob/e637ad972378619d580b6680786121d28065c241/librz/util/ubase64.c#L392-L416

wargio commented on librz/util/base32.c:
```diff
@@ -0,0 +1,356 @@
+	ut8 final_group[5] = { 0 };
+	size_t ret;
+	rz_return_val_if_fail(src, 0);
+	rz_return_val_if_fail(dest, 0);
```
as before join them and move to line 164

BigFish2086 replied:
updated in 9a4203e4c7

wargio commented on librz/util/base32.c:
```diff
@@ -0,0 +1,356 @@
+RZ_API RZ_OWN char *rz_base32_encode_dyn(RZ_NULLABLE const ut8 *src, size_t n) {
+	rz_return_val_if_fail(src, NULL);
+	size_t buf_sz = calculate_dest_length(n) + 1;
+	char *out = (char *)malloc(buf_sz);
```
as before, i don't see where `out[buf_sz-1]` is set to `0`

BigFish2086 replied:
updated in 34d4219689

wargio commented on librz/util/base32.c:
```diff
@@ -0,0 +1,356 @@
+	st64 ret = 0;
+
+	rz_return_val_if_fail(src, 0);
+	rz_return_val_if_fail(dest, 0);
```
join them and move them to line 255

BigFish2086 replied:
updated in 9a4203e4c7

wargio commented on librz/util/base32.c:
```diff
@@ -0,0 +1,356 @@
+ * \endcode
+ */
+RZ_API st64 rz_base32_decode(RZ_OUT RZ_NULLABLE ut8 *dest, RZ_NULLABLE const char *src, st64 n) {
+	char buf[8], tmp[5];
+	int c;
+	size_t i, j;
```
initialize these

BigFish2086 replied:
I think these variables are being intialized like: 

`i, j` are intialized in the loop
`c`    is having a different value every loop-iteration
`buf`  is being "populated" while going through the loop
`tmp`  is used as an output variable later in this line within the funciton call `unpack_from5`

but I updated them either way in 821d7e6710

wargio commented on librz/util/base32.c:
```diff
@@ -0,0 +1,356 @@
+		return ret;
+	} else if (j <= 1) {
+		return -1;
+	} else {
+		size_t rem = j;
+		for (; j < 8; j++) {
+			buf[j] = 'A'; // zero padding for partial group
+		}
+		unpack_from5((ut8 *)tmp, (const ut8 *)buf);
+		// decide how many of the unpacked bytes are valid based on how many base32 symbols you originally had.
+		static const int rem2bytes[8] = { 0, 0, 1, 0, 2, 3, 0, 4 };
+		int bytes = rem2bytes[rem];
+		if (bytes == 0) {
+			return -1;
+		}
+		memcpy(dest, tmp, bytes);
+		dest[bytes] = '\0';
+		return ret + bytes;
+	}
```
this else is useless. remove the else block and just do
```c
	if (j == 0) {
		dest[0] = '\0';
		return ret;
	} else if (j <= 1) {
		return -1;
	}
	
	size_t rem = j;
```

else blocks should be used only if there is other things to do after the `if-else` statement, but here we always return.

BigFish2086 replied:
updated in 47caf380fe

wargio commented on librz/util/base32.c:
```diff
@@ -0,0 +1,356 @@
+		return NULL;
+	}
+
+	st64 cap = 1 + 5 * ((len + 7) / 8);
```
add more parentheses to split addition from mult.

BigFish2086 replied:
updated in 47caf380fe

wargio commented on librz/util/base32.c:
```diff
@@ -0,0 +1,356 @@
+	/* NUL‑terminate and shrink to fit. */
+	buf[out_len] = '\0';
+	if (out_len + 1 < cap) {
+		ut8 *tmp = (ut8 *)realloc(buf, (size_t)out_len + 1);
+		if (tmp) {
+			buf = tmp;
+		}
```
as above, free buf and return null if realloc returns a null pointer.

wargio commented on librz/util/base36.c:
```diff
@@ -0,0 +1,113 @@
+RZ_API RZ_OWN char *rz_base36_encode_dyn(ut64 val) {
+	static const char alphabet[] = "0123456789abcdefghijklmnopqrstuvwxyz";
+
+	char tmp[RZ_BASE36_BUFSZ];
```
initialize this by doing `tmp[RZ_BASE36_BUFSZ] = { 0 };`

BigFish2086 replied:
updated in e762aeed4d

wargio commented on librz/util/base36.c:
```diff
@@ -0,0 +1,113 @@
+	size_t i;
+	// 64-bit base36 str has at most 13 characters
+	if (len > RZ_BASE36_BUFSZ) {
+		eprintf("Error: base36_decode supports up to 64-bit values only\n");
```
```suggestion
		RZ_LOG_ERROR("base36_decode supports up to 64-bit values only\n");
```
RZ_LOG_ERROR, do not use eprintf

BigFish2086 replied:
updated in f769d36c08

wargio commented on librz/util/base36.c:
```diff
@@ -0,0 +1,113 @@
+		// Check for overflow
+		if (i == 12) {
+			if (v > 3 || UT64_ADD_OVFCHK(ret, v * pow36[i])) {
+				eprintf("Error: base36_decode supports up to 64-bit values only\n");
```
```suggestion
				RZ_LOG_ERROR("base36_decode supports up to 64-bit values only\n");
```
same here.

BigFish2086 replied:
updated in f769d36c08

wargio commented on librz/util/base85.c:
```diff
@@ -30,142 +32,403 @@
+ * The function starts with the most‑significant decoded byte and appends up to
+ * four bytes to the buffer referenced by \p d.
+ */
+static void decode_tuple_buf(unsigned long tuple, int count, char **d, size_t *len) {
+	for (int i = 1; i < count; i++) {
```
is `int` actually ok for a size?

BigFish2086 replied:
I changed it to `size_t` in c10c152f0ff863242d181c0b4efeb13b3df5c7af - I thought it might be ok since it's used internally, but it's actually better to use `size_t` either way.

wargio commented on librz/util/base85.c:
```diff
@@ -30,142 +32,403 @@
+ * (not counting that NUL).  If either \p dest or \p src is <code>NULL</code>,
+ * nothing is written and the function returns 0.
+ */
+RZ_API int rz_base85_encode(char *dest, const char *src, size_t n, int delims, int wrap, int y_abbr) {
+	if (!dest || !src) {
+		return 0;
+	}
```
add `RZ_NULLABLE`

BigFish2086 replied:
updated in 942673555ffb240bcdbca971b6a422070bc08a60

wargio commented on librz/util/base85.c:
```diff
@@ -30,142 +32,403 @@
+RZ_API RZ_OWN char *rz_base85_encode_dyn(RZ_NULLABLE const char *src, size_t n, int delims, int wrap, int y_abbr) {
+	rz_return_val_if_fail(src, NULL);
+
+	char *ret = (char *)malloc(rz_base85_enc_buflen(n, delims, wrap));
```
`rz_base85_enc_buflen` compared to the other encoders returns always a len + delimiter? should this be unified as the others to return just the calculated size ?

BigFish2086 replied:
I think it's also fine like that, its signature points to it doing something with `delims` and line `wraps`. Also, one won't need to do the rest of the calculations every time before allocating the encoding output buffer, which they can miss or do it wrongly. What do you think?

wargio commented on librz/util/base85.c:
```diff
@@ -30,142 +32,403 @@
+		return NULL;
+	}
+	int out_len = rz_base85_encode(ret, src, n, delims, wrap, y_abbr);
+	(void)out_len;
```
```suggestion
	rz_base85_encode(ret, src, n, delims, wrap, y_abbr);
```

BigFish2086 replied:
updated in c87b35f9128859651236ef1e85030a5087f8fdab

wargio commented on librz/util/base85.c:
```diff
@@ -30,142 +32,403 @@
 			if (ignore_garbage) {
 				continue;
 			}
 			eprintf("ascii85: invalid character '%c'\n", c);
```
RZ_LOG_ERROR

BigFish2086 replied:
updated in b07a5a959e1b07d75236f07592150f594efb43ee

wargio commented on librz/util/base85.c:
```diff
@@ -30,142 +32,403 @@
-				eprintf("ascii85: ~ without >\n");
-				return false;
+			if (getc_nospace_buf(src, len, &pos) != '>') {
+				eprintf("ascii85: '~' not followed by '>'\n");
```
RZ_LOG_ERROR

BigFish2086 replied:
updated in b07a5a959e1b07d75236f07592150f594efb43ee

wargio commented on librz/util/base85.c:
```diff
@@ -30,142 +32,403 @@
-	return true;
+
+	if (delims && !have_end) {
+		eprintf("ascii85: missing ~>\n");
```
```suggestion
		RZ_LOG_ERROR("ascii85: missing ~>\n");
```

BigFish2086 replied:
updated in b07a5a959e1b07d75236f07592150f594efb43ee

wargio commented on librz/util/base85.c:
```diff
@@ -30,142 +32,403 @@
+
+	/* shrink‑to‑fit and NUL‑terminate */
+	if (written + 1 < (st64)cap) {
+		char *tmp = (char *)realloc(buf, (size_t)written + 1);
+		if (tmp) {
+			buf = tmp;
+		}
```
same here. if realloc fails, free buf and return null.

Rot127 replied:
Ping on this one.

BigFish2086 replied:
@wargio, @Rot127 please check this one about why not returning null while try shrinking the buffer: https://github.com/rizinorg/rizin/pull/5271/files#r2234010029

wargio commented on subprojects/rizin-shell-parser/src/parser.c:
what is this? why so many changes? @notxvilka

BigFish2086 replied:
my bad, I didn't actually notice that before making the PR, I did the rebase as mentioned in the [CONTRIBUTING.md](https://github.com/rizinorg/rizin/blob/dev/CONTRIBUTING.md): `git rebase -i dev`, though I don't think that I actually did any changes to the parser.c file, may be it's better to remove it from this PR.

wargio commented on librz/crypto/crypto.c:
```diff
@@ -41,6 +41,10 @@ static const struct {
+	{ "base85", RZ_CODEC_B85 },
+	{ "base36", RZ_CODEC_B36 },
+	{ "base32", RZ_CODEC_B32 },
+	{ "base16", RZ_CODEC_B16 },
```
maybe move punycode as last and sort them by base?

BigFish2086 replied:
updated in 21328b3f6245303d7c23463d89dd31cc06fb288e

wargio reviewed (changes requested):
many things to change, but why such big change in the parser?

notxvilka commented on librz/util/base16.c:
```diff
@@ -0,0 +1,290 @@
+// SPDX-FileCopyrightText: 2025 Ahmed Ibrahim <a.ibrahim8686@gmail.com>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+/**
```
Since base16 is essentially a hexdump, please share the code with `librz/util/hex.c` - `rz_hex_bin2str()` in particular.

BigFish2086 replied:
I don't really get what should be done in here :sweat_smile: So, can you elaborate more?

notxvilka replied:
Both should share the same code, just call that function instead.

notxvilka commented on librz/include/rz_util/rz_base85.h:
```diff
@@ -0,0 +1,20 @@
+RZ_API st64 rz_base85_decode(char *dest, const char *src, st64 len, int delims, int ignore_garbage);
+
+RZ_API RZ_OWN char *rz_base85_encode_dyn(const char *src, size_t n, int delims, int wrap, int y_abbr);
+RZ_API RZ_OWN char *rz_base85_decode_dyn(const char *src, st64 len, int delims, int ignore_garbage, size_t *out_len);
```
Please use the uniform API interface across all baseXX codecs.

BigFish2086 replied:
I think what you mean by the uniform API interface is to use `RZ_OUT` and `RZ_NULLABLE` and updated it like this in 942673555ffb240bcdbca971b6a422070bc08a60

notxvilka commented:
Yes, parser should not be changed

BigFish2086 committed:
Restore subprojects/rizin-shell-parser/parser.c to match origin/dev

notxvilka commented:
```
../test/unit/test_base85.c:50:15: error: variable 'newlines' set but not used [-Werror,-Wunused-but-set-variable]
   50 |         int col = 0, newlines = 0;
      |                      ^
1 error generated.
```

BigFish2086 committed:
Fix the order by baseXX

BigFish2086 committed:
Fix base16 - invert the logic in calculate_src_length

BigFish2086 committed:
Fix base16 - intialize variables in rz_base16_encode

BigFish2086 committed:
Fix base16 - null terminate the output buffer of rz_base16_encode_dyn

BigFish2086 committed:
Fix base16 - get rid of unnecessary else in rz_base16_decode

BigFish2086 committed:
Fix base16 - use `len & 1` instead of `(len % 2) != 0`

BigFish2086 committed:
Fix base32 - invert the logic in calculate_src_length

BigFish2086 committed:
Fix base32 - compress two return statments by using `rz_return_val_if_fail(src && dest, 0);`

BigFish2086 committed:
Fix base32 - null terminate the output of rz_base32_encode

BigFish2086 committed:
Fix base32 - get rid of unnecessary else in rz_base32_decode

add more parentheses to split addition from mult for better clarity

BigFish2086 committed:
Fix base32 - intialize variables in rz_base32_encode_dyn

BigFish2086 committed:
Fix base36 - intialize `tmp` variable in rz_base36_encode_dyn

BigFish2086 committed:
Fix base36 - use RZ_LOG_ERROR instead of eprintf

BigFish2086 committed:
Fix base36 - use RZ_NULLABLE for the API interface

BigFish2086 committed:
Fix base85 - use RZ_OUT & RZ_NULLABLE for the API interface

BigFish2086 committed:
Fix base85 - use RZ_LOG_ERROR instead of eprintf

BigFish2086 committed:
Fix base85 - use size_t instead of int for decode_tuple_buf

BigFish2086 committed:
Fix base85 - remove unused varaible, `out_len`, in rz_base85_encode_dyn

BigFish2086 committed:
Fix base85 - correct decode buffer size calculation

The previous rz_base85_dec_buflen() underestimated the worst‑case output
size (3 bytes per 4 input chars), causing overflows when using ‘z’/’y’
abbreviations. Update it to allocate 4 bytes per input character plus
one for the NUL terminator, eliminating heap-buffer-overflow errors.

BigFish2086 committed:
Fix formatting

BigFish2086 commented:
Hi @Rot127 I think I have fixed the address sanitizer issue via using a correct upper bound for the decoding buffer in 48a62c2e3c9fe3cea7be9f6236651b8eddce0fef - I have use this to build on myside: `meson -Db_sanitize=address,undefined build`

BigFish2086 committed:
Fix base85-test - use `newlines` variable to make sure line breaks were inserted as expected

notxvilka commented on test/unit/test_base85.c:
```diff
@@ -0,0 +1,123 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
```
`../test/unit/test_base85.c(8): fatal error C1083: Cannot open include file: 'unistd.h': No such file or directory`

notxvilka reviewed (changes requested):
Why some functions called `_dyn()` and others - not? Please add missing functions to those that do not have it - e.g. to base36. This is what I meant by "uniform".

Please address also the formatting issue:
```
test/unit/test_base85.c:60:62: error: code should be clang-formatted [-Wclang-format-violations]
        mu_assert_true(col <= 10, "final line exceeds wrap length");
                                                                    ^
```

notxvilka reviewed (changes requested):
You forgot to update this test also:
```
[XX]  81 ms   db/tools/rz_hash rz-hash -L
RZ_NOPLUGINS=1 /Users/runner/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -esearch.show_progress=false -eflirt.sigdb.load.home=false -N -qc '!rz-hash -L' =
-- stdout
--- expected
+++ actual
@@ -77,16 +77,20 @@
 ED____ rc6            LGPL-3     rakholiyajenish.07             RC6 symmetric-key block cipher
 _D____ ror            LGPL-3     pancake                        Rotate Right symmetric-key block cipher
 __ed__ base64         LGPL-3     rakholiyajenish.07             Base64 encoder/decoder
+__ed__ base32         LGPL-3     Ahmed Ibrahim                  base32 encoder/decoder
+ED____ rc2            LGPL-3     lionaneesh                     RC2 symmetric-key block cipher
 ED____ xor            LGPL-3     pancake                        XOR symmetric-key block cipher
-ED____ rc2            LGPL-3     lionaneesh                     RC2 symmetric-key block cipher
 ED____ cps2           LGPL-3     pancake,esanfelix,pof          Capcom Play System 2 (CPS-2) symmetric-key block cipher
 ED____ rot            LGPL-3     pancake                        Caesar symmetric-key cipher
 __ed__ punycode       LGPL-3     pancake                        Punycode encoder/decoder
 ED____ serpent-ecb    LGPL-3     NicsTr                         Serpent symmetric-key block cipher (ECB block mode)
 ED____ sm4-ecb        LGPL-3     0xSh4dy                        ShangMi 4 symmetric-key block cipher (ECB block mode)
+__ed__ base36         LGPL-3     abcSup                         base36 encoder/decoder
+__ed__ base85         LGPL-3     Remy Oukaour                   base85 encoder/decoder
 ED____ aes-cbc        LGPL-3     rakholiyajenish.07             AES symmetric-key block cipher (CBC block mode)
 E_____ rol            LGPL-3     pancake                        Rotate Left symmetric-key block cipher
 ED____ aes-ecb        LGPL3      Nettle project,pancake         AES symmetric-key block cipher (ECB block mode)
+__ed__ base16         LGPL-3     Ahmed Ibrahim                  base16 encoder/decoder
 ED____ des-ecb        LGPL-3     deroad                         DES symmetric-key block cipher (ECB block mode)
 ED____ blowfish       LGPL3      kishorbhat                     Blowfish symmetric-key block cipher
 __ed__ base91         LGPL-3     rakholiyajenish.07             Base91 encoder/decoder
```

notxvilka commented on librz/crypto/p/crypto_base36.c:
```diff
@@ -0,0 +1,105 @@
+
+	if (cry->dir == RZ_CRYPTO_DIR_ENCRYPT) {
+		if (len > 8) { /* > 64‑bit won’t fit in our encoder */
+			eprintf("base36: encoder supports up to 64‑bit values (%d > 8 bytes)\n", len);
```
`RZ_LOG_ERROR()`

notxvilka commented on librz/util/base16.c:
```diff
@@ -0,0 +1,286 @@
+	size_t i, j;
+	st64 ret = 0;
+
+	rz_return_val_if_fail(src, 0);
```
Move these to the top of the function

notxvilka commented:
@BigFish2086 hi, how is it going with addressing the remaining feedback?

BigFish2086 commented:
hi @notxvilka, sorry for being late. I will be working more on it in the upcoming days.

BigFish2086 committed:
Fix base85-test - remove unnecessary includes

BigFish2086 committed:
Fix rz-hash test - updated rz-hash -L expected result

BigFish2086 committed:
Fix crypto_base36 - use RZ_LOG_ERROR instead of eprintf

BigFish2086 committed:
Fix base16 - compress rz_return_val_if_fail statements and move before locals

BigFish2086 committed:
Update base16 - use hex.c existing encoding/decoding logic

- base16.c encoding/decoding functions are now wrappers for hex.c
  `rz_hex_bin2str` and `rz_hex_str2bin` functions
- updated related base16 unit testing and documentation
- updated the integraion with rz-hash binary (or crypto_base16)

BigFish2086 committed:
Fix formatting

notxvilka commented on librz/util/base16.c:
```diff
@@ -0,0 +1,144 @@
+	if (!out) {
+		return NULL;
+	}
+	(void)rz_base16_encode(out, src, n);
```
Return NULL if return value is 0

notxvilka commented on librz/util/base32.c:
```diff
@@ -0,0 +1,352 @@
+	if (!out) {
+		return NULL;
+	}
+	(void)rz_base32_encode(out, src, n);
```
Same here, return NULL if return value is wrong

notxvilka commented on librz/util/base16.c:
```diff
@@ -0,0 +1,144 @@
+ */
+RZ_API int rz_base16_encode(RZ_OUT RZ_NULLABLE char *dest, RZ_NULLABLE const ut8 *src, size_t n) {
+	rz_return_val_if_fail(src && dest, 0);
+	rz_hex_bin2str(src, (int)n, dest);
```
Check the return value of `rz_hex_bin2str()`

notxvilka replied:
Could you please also add a Doxygen to the `rz_hex_bin2str()` (in `librz/util/hex.c`) while you are on it? For some reason it is not documented while `rz_hex_str2bin()` is.

notxvilka reviewed (changes requested):
Looks better now. Few remaining issues left.

notxvilka commented:
Don't forget to rebase as well, please.

BigFish2086 commented:
regarding this one:
> Why some functions called `_dyn()` and others - not? Please add missing functions to those that do not have it - e.g. to base36. This is what I meant by "uniform".

I actually realized that the result of the decoding should have been `ut64` as it was in `iob_net.c` (which is another error on my side :sweat_smile:) , so I wanted to ask about how to approach it - to keep the api like others as well

```c
// this would encode the ut64 value into bout, returning the length of the encoded string
RZ_API size_t rz_base36_encode(RZ_OUT RZ_NULLABLE char *bout, ut64 val);

// this would encode the ut64 value and return the result in a newlly allocated buffer
RZ_API RZ_OWN char *rz_base36_encode_dyn(ut64 val);

// this now decodes into bout - a ut64 as before and return the state through st64
// and not bool to match other APIs, so zero on success and negative on errors
RZ_API st64 rz_base36_decode(RZ_OUT RZ_NULLABLE ut64 *bout, RZ_NULLABLE const char *bin, st64 len);

// this should now use rz_base36_decode internally, so it would know about the failure if any, but how to report it
// since it should return ut64?
// - can have it return ut64* and it will be NULL if any errors (which is aligned with most apis, but they're returning ut8*)
// - can have it return ut64 and accept another bool* flage, so the user can check it for errors
RZ_API st64 rz_base36_decode_dyn(RZ_NULLABLE const char *bin, const size_t len);
```

so, what is better in here?

notxvilka commented:
> regarding this one:
> 
> > Why some functions called `_dyn()` and others - not? Please add missing functions to those that do not have it - e.g. to base36. This is what I meant by "uniform".
> 
> I actually realized that the result of the decoding should have been `ut64` as it was in `iob_net.c` (which is another error on my side 😅) , so I wanted to ask about how to approach it - to keep the api like others as well
> 
> ```c
> // this would encode the ut64 value into bout, returning the length of the encoded string
> RZ_API size_t rz_base36_encode(RZ_OUT RZ_NULLABLE char *bout, ut64 val);
> 
> // this would encode the ut64 value and return the result in a newlly allocated buffer
> RZ_API RZ_OWN char *rz_base36_encode_dyn(ut64 val);
> 
> // this now decodes into bout - a ut64 as before and return the state through st64
> // and not bool to match other APIs, so zero on success and negative on errors
> RZ_API st64 rz_base36_decode(RZ_OUT RZ_NULLABLE ut64 *bout, RZ_NULLABLE const char *bin, st64 len);
> 
> // this should now use rz_base36_decode internally, so it would know about the failure if any, but how to report it
> // since it should return ut64?
> // - can have it return ut64* and it will be NULL if any errors (which is aligned with most apis, but they're returning ut8*)
> // - can have it return ut64 and accept another bool* flage, so the user can check it for errors
> RZ_API st64 rz_base36_decode_dyn(RZ_NULLABLE const char *bin, const size_t len);
> ```
> 
> so, what is better in here?

I think just casting should be sufficient.

wargio reviewed (approved):
i'm ok with the changes, just a small nitpic that doesn't need to be corrected: all the `rz_base*_dyn` functions has all different args, wouldn't be better to have all the same args?

Rot127 commented:
Just triggered the `dist-` CI because there is endian dependent logic in there.

Rot127 commented on librz/util/base85.c:
```diff
@@ -30,142 +32,402 @@
+ * (not counting that NUL).  If either \p dest or \p src is <code>NULL</code>,
+ * nothing is written and the function returns 0.
+ */
+RZ_API int rz_base85_encode(RZ_OUT RZ_NULLABLE char *dest, RZ_NULLABLE const char *src, size_t n, int delims, int wrap, int y_abbr) {
```
`\return` is missing in doxygen.

Rot127 replied:
Also, shouldn't `dest` and `src` be `RZ_NONNULL` and tested with `rz_return_val_if_fail`?

Rot127 commented on librz/util/base85.c:
```diff
@@ -30,142 +32,402 @@
+ *
+ * The caller becomes owner of the returned buffer and must <tt>free()</tt> it.
+ */
+RZ_API RZ_OWN char *rz_base85_encode_dyn(RZ_NULLABLE const char *src, size_t n, int delims, int wrap, int y_abbr) {
```
```suggestion
RZ_API RZ_OWN char *rz_base85_encode_dyn(RZ_NONNULL const char *src, size_t n, int delims, int wrap, int y_abbr) {
```

Rot127 commented on librz/util/base85.c:
```diff
@@ -30,142 +32,402 @@
+ *         the Ascii 85 representation, or <code>NULL</code> if either
+ *         <code>src == NULL</code> or a memory allocation fails.
+ *
+ * The caller becomes owner of the returned buffer and must <tt>free()</tt> it.
```
```suggestion
```

Already implied by `RZ_OWN`

Rot127 commented on librz/util/base85.c:
```diff
@@ -30,142 +32,402 @@
+ * \attention The caller must allocate \p dest large enough.  A safe upper bound is
+ * \c 1 + 3 × ((len + 1)/4) bytes.
+ */
+RZ_API st64 rz_base85_decode(RZ_OUT RZ_NULLABLE char *dest, RZ_NULLABLE const char *src, st64 len, int delims, int ignore_garbage) {
```
```suggestion
RZ_API st64 rz_base85_decode(RZ_OUT RZ_NONNULL char *dest, RZ_NONNULL const char *src, st64 len, int delims, int ignore_garbage) {
```

Rot127 commented on librz/util/base85.c:
```diff
@@ -30,142 +32,402 @@
+ * \return Pointer to a newly allocated buffer containing the raw bytes,
+ *         or \c NULL on error.  The buffer is NUL‑terminated for convenience.
+ */
+RZ_API RZ_OWN char *rz_base85_decode_dyn(RZ_NULLABLE const char *src, st64 len, int delims, int ignore_garbage, size_t *out_len) {
```
```suggestion
RZ_API RZ_OWN char *rz_base85_decode_dyn(RZ_NONNULL const char *src, st64 len, int delims, int ignore_garbage, size_t *out_len) {
```

Rot127 commented on librz/include/rz_util/rz_base85.h:
```diff
@@ -0,0 +1,20 @@
+extern "C" {
+#endif
+
+RZ_API int rz_base85_encode(RZ_OUT RZ_NULLABLE char *dest, RZ_NULLABLE const char *src, size_t n, int delims, int wrap, int y_abbr);
+RZ_API st64 rz_base85_decode(RZ_OUT RZ_NULLABLE char *dest, RZ_NULLABLE const char *src, st64 len, int delims, int ignore_garbage);
+
+RZ_API RZ_OWN char *rz_base85_encode_dyn(RZ_NULLABLE const char *src, size_t n, int delims, int wrap, int y_abbr);
+RZ_API RZ_OWN char *rz_base85_decode_dyn(RZ_NULLABLE const char *src, st64 len, int delims, int ignore_garbage, size_t *out_len);
```
```suggestion
RZ_API int rz_base85_encode(RZ_OUT RZ_NONNULL char *dest, RZ_NONNULL const char *src, size_t n, int delims, int wrap, int y_abbr);
RZ_API st64 rz_base85_decode(RZ_OUT RZ_NONNULL char *dest, RZ_NONNULL const char *src, st64 len, int delims, int ignore_garbage);

RZ_API RZ_OWN char *rz_base85_encode_dyn(RZ_NONNULL const char *src, size_t n, int delims, int wrap, int y_abbr);
RZ_API RZ_OWN char *rz_base85_decode_dyn(RZ_NONNULL const char *src, st64 len, int delims, int ignore_garbage, size_t *out_len);
```

Rot127 commented on librz/util/base16.c:
```diff
@@ -0,0 +1,144 @@
+ * This function converts each input byte to two lowercase hexadecimal
+ * characters using \c rz_hex_bin2str().
+ */
+RZ_API int rz_base16_encode(RZ_OUT RZ_NULLABLE char *dest, RZ_NULLABLE const ut8 *src, size_t n) {
```
```suggestion
RZ_API int rz_base16_encode(RZ_OUT RZ_NONNULL char *dest, RZ_NONNULL const ut8 *src, size_t n) {
```

Rot127 commented on librz/util/base16.c:
```diff
@@ -0,0 +1,144 @@
+ *
+ * Allocates exactly \c (2 × n) + 1 bytes, encodes the input, and NUL-terminates.
+ */
+RZ_API RZ_OWN char *rz_base16_encode_dyn(RZ_NULLABLE const ut8 *src, size_t n) {
```
```suggestion
RZ_API RZ_OWN char *rz_base16_encode_dyn(RZ_NONNULL const ut8 *src, size_t n) {
```

Rot127 commented on librz/util/base16.c:
```diff
@@ -0,0 +1,144 @@
+ * This is a thin wrapper around \c rz_hex_str2bin().
+ * Output is **not** automatically NUL-terminated unless you add it yourself.
+ */
+RZ_API int rz_base16_decode(RZ_OUT RZ_NULLABLE ut8 *dest, RZ_NULLABLE const char *src) {
```
```suggestion
RZ_API int rz_base16_decode(RZ_OUT RZ_NONNULL ut8 *dest, RZ_NONNULL const char *src) {
```

Rot127 commented on librz/util/base16.c:
```diff
@@ -0,0 +1,144 @@
+ * \return Pointer to a newly allocated buffer containing the decoded
+ *         binary data followed by a NUL byte, or \c NULL on error.
+ *
+ * Accepts even or odd numbers of hex digits. For odd lengths, the final
+ * nibble is padded with zero. The returned buffer is always NUL-terminated.
+ * Caller must \c free() the returned pointer.
+ */
+RZ_API RZ_OWN ut8 *rz_base16_decode_dyn(RZ_NULLABLE const char *src, st64 len) {
+	rz_return_val_if_fail(src, NULL);
```
```suggestion
 * Accepts even or odd numbers of hex digits. For odd lengths, the final
 * nibble is padded with zero.
 */
RZ_API RZ_OWN ut8 *rz_base16_decode_dyn(RZ_NONNULL const char *src, st64 len) {
	rz_return_val_if_fail(src, NULL);
```

Rot127 commented on librz/util/base32.c:
```diff
@@ -0,0 +1,352 @@
+ *
+ * Any other value yields \c false.
+ */
+static bool is_base32(int c) {
```
```suggestion
static inline bool is_base32(int c) {
```

Rot127 commented on librz/util/base32.c:
```diff
@@ -0,0 +1,352 @@
+ * free(enc);
+ * \endcode
+ */
+RZ_API size_t rz_base32_encode(RZ_OUT RZ_NULLABLE char *dest, RZ_NULLABLE const ut8 *src, size_t n) {
```
```suggestion
RZ_API size_t rz_base32_encode(RZ_OUT RZ_NONNULL char *dest, RZ_NONNULL const ut8 *src, size_t n) {
```

Rot127 commented on librz/util/base32.c:
```diff
@@ -0,0 +1,352 @@
+ * assert(memcmp(buf, "foo", 3) == 0);
+ * \endcode
+ */
+RZ_API st64 rz_base32_decode(RZ_OUT RZ_NULLABLE ut8 *dest, RZ_NULLABLE const char *src, st64 n) {
```
```suggestion
RZ_API st64 rz_base32_decode(RZ_OUT RZ_NONNULL ut8 *dest, RZ_NONNULL const char *src, st64 n) {
```

Rot127 commented on librz/util/base36.c:
```diff
@@ -0,0 +1,112 @@
+ * \param val  The unsigned 64‑bit integer to encode.
+ * \return     Pointer to a freshly allocated, NUL‑terminated digit string,
+ *             or \c NULL if memory allocation fails.
+ *
+ * The caller \b must free the returned buffer with \c free().
+ *
+ */
```
```suggestion
 */
```

Rot127 commented on librz/util/base36.c:
```diff
@@ -0,0 +1,112 @@
+ * explicit overflow check: the digit must be ≤ 3 and the addition
+ * <code>ret + v × pow36[12]</code> must not wrap.
+ */
+RZ_API st64 rz_base36_decode(RZ_NULLABLE const char *str, const size_t len) {
```
```suggestion
RZ_API st64 rz_base36_decode(RZ_NONNULL const char *str, const size_t len) {
rz_return_va_if_fail(str && len > 13, -1);
```

Rot127 commented on librz/include/rz_util/rz_base85.h:
SPDX header mising

Rot127 commented on librz/include/rz_util/rz_base36.h:
SPDX header missing

Rot127 commented on librz/include/rz_util/rz_base32.h:
SPDX

Rot127 commented on librz/include/rz_util/rz_base16.h:
SPDX

Rot127 commented on librz/crypto/p/crypto_base85.c:
```diff
@@ -0,0 +1,72 @@
+
+RzCryptoPlugin rz_crypto_plugin_base85 = {
+	.name = "base85",
+	.author = "Remy Oukaour",
```
```suggestion
	.author = "Ahmed Ibrahim",
```

You changed pretty much everything.

Rot127 reviewed (changes requested):
Looks slick!
Thanks for all the work!

Please address the comments and it looks pretty much done.

BigFish2086 committed:
Add documentation to rz_hex_bin2str

BigFish2086 committed:
Fix base85 - add missing `reutrn` docs to rz_base85_encode function

BigFish2086 committed:
Fix base85 - use RZ_NONNULL for dest and src function instead of RZ_NULLABLE

BigFish2086 committed:
Fix base36 - use RZ_NONNULL for dest and src function instead of RZ_NULLABLE

BigFish2086 committed:
Fix base16 - use RZ_NONNULL for dest and src function instead of RZ_NULLABLE

BigFish2086 committed:
Fix base32 - use RZ_NONNULL for dest and src function instead of RZ_NULLABLE

BigFish2086 committed:
Fix base16 - add missig checks for invalid inputs in rz_base16_encode & rz_base16_encode_dyn

BigFish2086 committed:
Fix base32 - add missing checks for bad encoding

BigFish2086 committed:
Fix - add missing SPDX

notxvilka commented:
@BigFish2086 please address all our comments and it will be in a good enough shape to merge, I think.

BigFish2086 committed:
Fix Formatting

BigFish2086 committed:
Add rz_base36_encode & rz_base36_decode_dyn functions to base36 API
